[Document Settings]: Add summary panel version 1

ntsekouras opened the pull request:
Part of: https://github.com/WordPress/gutenberg/issues/39082

This PR intends to add a fist iteration of `post summary` panel in inspector controls. The new post summary panel will include `Featured Image, post title, excerpt and author`. All of the controls except `post title` which didn't exist before in inspector controls were removed from their previous positions/panels.

## Notes
1. Previously `post excerpt and featured image` would live in their own `panel` and their visibility was also controlled by the `preferences -> panels` section. I **have not removed** the preferences, and we display these conditionally on the new panel, based on these preferences.
2. Do you think there is a risk of breaking backwards compatibility by removing the `Panels` in `edit-post` package for the above(excerpt and featured image)? I think not because they are not exposed anywhere as I could see..
3. In the exported PostFeaturedImage component from `editor` package [two wrappers have been added](https://github.com/WordPress/gutenberg/pull/39973/files#diff-9b8c234ca3c82c5dcc1841899d5825553a4fd559b4c7d03c96a827b0ecbff314R127). Do you think this can break backwards compatibility by maybe some third party targeting internal elements with selectors or something similar? 🤔

ntsekouras committed:
[Document Settings]: Add summary panel version 1

jasmussen committed:
Polish featured image.

jasmussen committed:
Tweak excerpt.

jasmussen commented:
Alright, based on work by @javierarce I've pushed some changes, and I think we're fast approaching something. Here's what it looks like now:

![status](https://user-images.githubusercontent.com/1204802/161261506-7b6eaaf2-6f0a-4a93-93a7-eaba31537630.gif)

Not bad for just a little work!

Thoughts:

* What is necessary for us to retire the separate Excerpt and Featured Image panels?
* How about the components (right now with big warnings in the code we are using 2 just for testing, but what's actually needed?)
* Same with the "Author" dropdown which exists inside Status & visibility, maybe we can remove it there now?
* Right now we are loading a TINY thumbnail image into the featured image placeholder. Can we load a bigger source? Does it have to be inside a responsive wrapper? (I think I'll take a stab at removing it, see what happens there, that would also allow us to let it go edge to edge).
* The author dropdown needs a little love to be simplified. What are our options here? It might be simplest to keep it a dropdown for now, since there's some specialized behavior here, but are there any visual properties we can leverage, such as `isSmall` or otherwise?
* The PostTitle component in the sidebar just working is 👌 — but unfortunately it can't be a `h1`. In fact I think it needs to be either a `h3` or a `p` or `div`, so as to not mess with the hierarchy. We'll want to apply some custom size styles to it regardless. 
* As noted in the code, there's a lot of junk we can remove if we replace the existing featured image and excerpt components instead of copying them. 

I'll dive in, but even just as is, this is feeling sweet to me. What do you think, Javi? And thank you Nik for the boost here, looking forward to your thoughts.

jasmussen commented:
It looks like we can remove the `ResponsiveWrapper` and load a higher res featured image version:
<img width="385" alt="Screenshot 2022-04-01 at 14 25 49" src="https://user-images.githubusercontent.com/1204802/161262997-5f10c017-0a39-4d2b-82be-fd87329fd657.png">

I haven't pushed that code because I'm not quite sure what it was there for. @ntsekouras any thoughts?

jasmussen committed:
Fix focus styles for edit and trash.

jasmussen committed:
Fix focus styles.

ntsekouras commented:
Just some early notes..

> What is necessary for us to retire the separate Excerpt and Featured Image panels?

This is might not be needed depending on the changes we are going to make and if we want to apply these to the existing ones. For `Excerpt` I think this won't be the case, but for `Featured Image` might be, especially if we only change the `remove/replace` buttons.

>Same with the "Author" dropdown which exists inside Status & visibility, maybe we can remove it there now?

Noting that the Author component changes from `select` to `combobox` depending on the number of users. A quick way to test this is by [changing this value](https://github.com/WordPress/gutenberg/blob/trunk/packages/editor/src/components/post-author/index.js#L14), although the label will not be at the side for this one..


>It looks like we can remove the ResponsiveWrapper and load a higher res featured image version:

I'll have to check the code better for this one.

javierarce commented:
@jasmussen I just took it for a spin it's working great! Very promising!

jasmussen commented:
> This is might not be needed depending on the changes we are going to make and if we want to apply these to the existing ones. For Excerpt I think this won't be the case, but for Featured Image might be, especially if we only change the remove/replace buttons.

Sounds good, yeah I think the changes made to featured image for this branch are worth it even for the existing component.

We could potentially change the excerpt version to "minimal-excerpt" or something like that. But I still think we should remove the panel from the sidebar once this is integrated.

> Noting that the Author component changes from select to combobox depending on the number of users. A quick way to test this is by [changing this value](https://github.com/WordPress/gutenberg/blob/trunk/packages/editor/src/components/post-author/index.js#L14), although the label will not be at the side for this one..

Yes, we'll probably have to keep the dropdown mostly as is, potentially follow up separately. I meant, remove it from the Status & visibility panel so that it isn't duplicated at least.

jameskoster commented:
This is cool! :) 

One quirk I noticed is how the layout jumps when the featured image loads in. Could we display a placeholder with a Spinner to make that a bit less jarring?

Also, perhaps not something to do here, but would it make sense to put the permalink in the summary too?

<img width="400" src="https://user-images.githubusercontent.com/846565/161510106-cafe4f1f-026e-4e72-8b99-adc5716d0d15.gif" />

jasmussen commented:
This could be a good place for the permalink, probably yes. But I'd like to see if we can land this one in the next day or two 😅 

Agreed on the jump, I'm currently tinkering with a better thumbnail, let me see what I can do in terms of a base height.

jasmussen commented:
For reference to the conversation around why the featured image is wrapped in a `ResponsiveWrapper`, it appears to date back to [this conversation](https://github.com/WordPress/gutenberg/pull/894#discussion_r118701979) and [this PR](https://github.com/WordPress/gutenberg/pull/911):

> The idea is that after the first of these finishes loading, we know the dimensions of the image and the space we expect it to consume, even before the image itself has finished loading.

So it seems that wrapper is intended to help cover the jumping issue. Let me see if that's still the best way, or if there's something else we can do.

jasmussen committed:
Improve thumbnail resolution and appearance.

jasmussen commented:
Alright, I pushed some tweaks to the thumbnail handling. I ended up keeping the ResponsiveWrapper, as it did come with the benefit of not oversizing a small image, but still giving a max-width to a large image. I also set the desired size to `large`, with `thumbnail` as fallback, so there's a nice resolution when available. Current status:
![status](https://user-images.githubusercontent.com/1204802/161517784-29346cfe-7f21-4868-b8ac-d4d8364f6314.gif)

I also fixed the spinner position:

<img width="382" alt="Screenshot 2022-04-04 at 11 22 08" src="https://user-images.githubusercontent.com/1204802/161517836-e41e22c7-112c-4a02-a63e-056321a1395c.png">

The issue that Jay reported is one I could use advice on how to fix. Essentially this is the markup when the panel is fully loaded:
<img width="495" alt="Screenshot 2022-04-04 at 11 41 11" src="https://user-images.githubusercontent.com/1204802/161517987-b2817b40-992e-44b3-b6ce-6b12832ce3d1.png">
But this is the markup _while_ it's still loading:
<img width="552" alt="Screenshot 2022-04-04 at 11 41 40" src="https://user-images.githubusercontent.com/1204802/161518059-e58644f8-4968-4d2f-8cc9-fd62a026ba68.png">

As you can see, the entire `editor-post-featured-image` element is missing while the page is loading. I'm not entirely sure why that is. It would be a nice bonus to fix this issue, but since it's only on initial load, and is an existing problem with the featured image panel, I think it could be a nice bugfix to follow up on.

So in terms of next steps, here's the todo list:

* Rename `PostExcerpt2` to something better, since it appears we'll want to keep two copies of this one. Right?
* Remove `PostFeaturedImage` and replace with this new version, which is an improvement to the existing component rather than something that needs to be separate.
* Remove redundant code as part of that (see @todo's)
* Remove the duplicate "Author" from Status & Visibility panel
* Remove the duplicate Featured Image panel
* Remove the duplicate Excerpt panel
* `PostTitle` is hardcoded to output an H1, and it needs to be either a H3, P or a SPAN in the inspector. Can we add the element as a prop and send a SPAN in the inspector?

How does that sound, @ntsekouras? Do you think we could address and land those in the next couple of days?

ntsekouras commented:
>How does that sound, @ntsekouras? Do you think we could address and land those in the next couple of days?

Thanks for all the work here @jasmussen. I'll start working on this right now so hopefully we'll be able to land the v1.

ntsekouras committed:
update FeaturedImage in inspector controls + remove the previous panel.

The preference for displaying the previous panel is still respected and will show/hide the featured image from `post summary` now.

ntsekouras committed:
add a `minimal` version of PostSummary editor component and remove the `excerpt panel` from inspector controls

ntsekouras committed:
remove obsolete export

ntsekouras committed:
fix postExcerpt callback

ntsekouras committed:
remove obsolete styles import

ntsekouras committed:
remove PostAuthor from status panel

ntsekouras committed:
make post title a `span`

ntsekouras commented on packages/edit-post/src/components/sidebar/post-title/style.scss:
```diff
@@ -0,0 +1,3 @@
+.edit-post-post-title {
+	font-size: 2em;
```
@jasmussen this is a `span` for the post title, that will need the desired styles 😄

jasmussen replied:
Excellent let me take a stab in a minute, thank you!

jasmussen committed:
Polish heading.

jasmussen commented:
<img width="1392" alt="Screenshot 2022-04-04 at 17 50 05" src="https://user-images.githubusercontent.com/1204802/161582710-4100b44c-2bd6-4c18-8995-9905727e35d6.png">

This is feeling pretty good to me. @javierarce, @kjellr, @jameskoster can you take a quick look and see if this isn't something we can land?

jasmussen commented:
I took the liberty of marking this one as ready for review 😅 — we can always go back on that if need be. But I think the unification of several panels into one colophon is worth trying to land!

ntsekouras commented:
Failing tests are related and will update them first thing tomorrow.

ntsekouras committed:
fix tests and add a new one

javierarce commented:
This works great! 

The only thing I'd adjust is the white space around the title, I think there's too much of it. And I also found this bug when I added a link: the control appears behind the sidebar.


<img width="494" alt="Screenshot 2022-04-05 at 09 21 10" src="https://user-images.githubusercontent.com/4933/161700127-596d6c77-292c-4f87-9303-fc9d86e896b4.png">

Finally, I think we should give some affordance/hover states to the title, the excerpt, and the featured image. This is especially important when using the keyboard to select the fields.

jasmussen commented:
> And I also found this bug when I added a link: the control appears behind the sidebar.

Can you test that in trunk as well? I think that's an existing issue. 

I'll take a look at your feedback!

jasmussen committed:
Spacing and focus styles.

jasmussen commented:
Alright @javierarce how about this?
![status](https://user-images.githubusercontent.com/1204802/161705468-d373f4a3-386d-4a20-8978-3b1d71919ee9.gif)

javierarce commented:
> Alright @javierarce how about this?

It looks great, thanks!

> Can you test that in trunk as well? I think that's an existing issue.

Oh, you are right, and I noticed that in our case the Link Control gets open in the canvas too, so it stays behind the sidebar. I'll try to find or open an issue for that then.

jasmussen commented:
I think #38893 intended to fix it, but as it turns out it's non-trivial.

jasmussen commented:
Oh, and may we have a green check on the design side? 😇

javierarce reviewed (approved):
> Oh, and may we have a green check on the design side? 😇

Yes! 🚀

jasmussen commented:
@Mamaduka is there any chance I could get your eyes and instincts on this one?

I'd love to land it, ideally soon, as I think there's a big benefit to it. It works well, and if we run into anything, a revert is always an option.

Mamaduka commented:
Sure thing 👍  I am adding to my to-do list. I will try to review it today or tomorrow.

jameskoster commented:
Another quick thought on this one; should the excerpt wrapper have a max-height and scroll behaviour for overflows? I know excerpts by nature should be short, but short is relative.

jasmussen commented:
> Another quick thought on this one; should the excerpt wrapper have a max-height and scroll behaviour for overflows? I know excerpts by nature should be short, but short is relative.

I think that can be a perfect followup PR perhaps?

Mamaduka commented:
The feature looks great. Excellent work 🥇 

I noticed a few issues that we can address in follow-up PRs.

Upload spinner is displayed after a few seconds, and it's not clear that the Featured Image started uploading. Here's a screencast with Network throttling (Fast 3G); it's also visible with a normal connection.

https://user-images.githubusercontent.com/240569/161794663-ae630ae2-c5aa-4bf0-8cd4-06d6c0bc7f5c.mp4

The `title` and `author` rows are missing post-type support checks. Also, what happens if post-type doesn't support any of these fields?

This CPT only supports the editor field.

![CleanShot 2022-04-05 at 19 32 59](https://user-images.githubusercontent.com/240569/161795179-6150a6ad-ddc2-49c0-b501-12ce7e685b78.png)

jasmussen commented:
Thanks a ton for the review. I'll let this sit overnight and then see if we can't press the button tomorrow and follow up on all your points.

jasmussen commented:
@ntsekouras would you mind if we land this one and look at followups? I think it's ready, and we'll have some time in beta to follow up on any bugs noted on this thread.

ntsekouras committed:
add post-type support checks for title and author

ntsekouras commented:
>The title and author rows are missing post-type support checks. Also, what happens if post-type doesn't support any of these fields?

Good points @Mamaduka! I added the checks and in the case of nothing supported, I think it's okay to look into it in a follow up.

Mamaduka commented:
Sounds good, Nik.

The feature image issue exists on the trunk as well. So it's not related to this PR, but it would be a good improvement.
